### PR TITLE
DelayFlow with flexible DelayStrategy implementation #89

### PR DESCRIPTION
**Pull request by manonthegithub**
Implementation of flexible DelayStage with ability to create your own delay strategies with different delay policies. 

Refs #89

### CODE REVIEW

File: contrib/src/main/scala/akka/stream/contrib/DelayFlow.scala [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    override def onPull(): Unit = {
+      pull(in)
+    }
+
+    override def onTimer(timerKey: Any): Unit = {
+      push(out, delayedElem)
+      if (isClosed(in)) {
+        super.onUpstreamFinish()
```
Comment (patriknw) [inline contrib/src/main/scala/akka/stream/contrib/DelayFlow.scala]: `completeStage()` is probably more correct

**Comment by manonthegithub** [comment]
@patriknw made improvements according to your comments

**Comment by manonthegithub** [comment]
rebased PR to resolve conflict

**Comment by manonthegithub** [comment]
Hey guys. There was one fail in  that is not related to my PR: 
`[info] TimeWindow flow
[info] - should aggregate data for predefined amount of time *** FAILED *** (634 milliseconds)`.

Also there were fails in fixed delay test. The problem is that message comes to `expectNoMsg()` before its timer triggers. What is timer precision? How long should be the gap for test to pass for sure?

File: contrib/src/test/scala/akka/stream/contrib/DelayFlowSpec.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      val elems = Vector(1, 2, 3, 4, 5, 6, 7, 8, 9, 0)
+
+      val probe = Source(elems)
+        .via(DelayFlow[Int](fixedDelay))
+        .runWith(TestSink.probe)
+        .request(elems.size)
+
+      elems.foreach(probe
+        //the timer is not precise enough and fails occasionally if exact delay value is used
+        .expectNoMsg(fixedDelay - (10 millis))
```
Comment (manonthegithub) [inline contrib/src/test/scala/akka/stream/contrib/DelayFlowSpec.scala]: @patriknw @johanandren 
This test works on my machine, but failed on build server.
Can you guys please advice the size of the gap for test to pass for sure?

Comment (arpanchaudhury) [inline contrib/src/test/scala/akka/stream/contrib/DelayFlowSpec.scala]: You can run **validatePullRequest** task as mentioned in the contribution guidelines (if available for this project).

https://github.com/akka/akka/blob/master/CONTRIBUTING.md

Comment (johanandren) [inline contrib/src/test/scala/akka/stream/contrib/DelayFlowSpec.scala]: Hmm, that the ci-servers are slow should never cause a message to be emitted faster than the delay, right? Is it really the `expectNoMsg` failing?

Comment (manonthegithub) [inline contrib/src/test/scala/akka/stream/contrib/DelayFlowSpec.scala]: I guess scheduler in TimerGraphStageLogic is not precise enough... can it be the case?

Comment (johanandren) [inline contrib/src/test/scala/akka/stream/contrib/DelayFlowSpec.scala]: It's the actor system scheduler, which can trigger late, but it should never be early. The general thing with the CI-servers is rather that they are slow, so things may take more time than expected. I think shorter times may indicate a bug lurking.

Comment (manonthegithub) [inline contrib/src/test/scala/akka/stream/contrib/DelayFlowSpec.scala]: I'll try find out what's wrong, but I was unable to reproduce it despite some attempts today.

**Comment by manonthegithub** [comment]
I guess it could be working wrong because of dilated, also made `request()` closer to `expectNoMsg`

**Comment by johanandren** [comment]
Now the last run failed just because I gave a too early LGTM :)

**Comment by johanandren** [comment]
Please look into that failure (I see there was two other tests that also failed, so likely a slowness/timing failure, but still good to see if the test can be made not fail)

**Comment by manonthegithub** [comment]
Ok, I will investigate it more.

**Comment by manonthegithub** [comment]
Hey guys, I've got interesting results. I've corrected the test little bit:
```scala
"work with fixed delay" in {
      for(_ <- 1 to 1000) {
        val fixedDelay = 1 second

        val elems = Vector(1, 2)

        val probe = Source(elems)
          .log("l1", e => println(s"Test in $e: ${System.currentTimeMillis()}"))
          .via(DelayFlow[Int](fixedDelay.dilated))
          .log("l2", e => println(s"Test out $e: ${System.currentTimeMillis()}"))
          .runWith(TestSink.probe)

        elems.foreach(probe
          .request(1)
          .expectNoMsg(fixedDelay)
          .expectNext(_))

        probe.expectComplete()
      }

    }
```
And got interesting results in console:
```
.......
Test out 2: 1495794585819
Test in 1: 1495794585835
Test out 1: 1495794586866
Test in 2: 1495794586867
Test out 2: 1495794587870
Test in 1: 1495794587884
Test out 1: 1495794588929

assertion failed: received unexpected message OnNext(1)
java.lang.AssertionError: assertion failed: received unexpected message OnNext(1)
	at scala.Predef$.assert(Predef.scala:219)
	at akka.testkit.TestKitBase.expectNoMsg_internal(TestKit.scala:616)
	at akka.testkit.TestKitBase.expectNoMsg(TestKit.scala:612)
	at akka.testkit.TestKitBase.expectNoMsg$(TestKit.scala:612)
	at akka.testkit.TestKit.expectNoMsg(TestKit.scala:814)
	at akka.stream.testkit.TestSubscriber$ManualProbe.expectNoMsg(StreamTestKit.scala:550)
	at akka.stream.contrib.DelayFlowSpec.$anonfun$$init$$7(DelayFlowSpec.scala:42)
	at akka.stream.contrib.DelayFlowSpec.$anonfun$$init$$7$adapted(DelayFlowSpec.scala:43)

```

How can you explain this? What's wrong with test? @patriknw @johanandren

**Comment by johanandren** [comment]
The scheduler cannot guarantee exactly 1000 ms, it depends on a big amount of parameters (gc, jvm, threads, how the load on your machine is), 45ms delay sounds a bit much but the next element is delayed 3ms so that may be random. Do a benchmark of many more elements, say 10 000 and look at the average and the min/max. Also, don't use `System.currentTimeMillis` which usually have a higher granularity than 1ms, use `System.nanoTime` instead.

**Comment by manonthegithub** [comment]
@johanandren Yeah, I will do more sample measurements. But I am about test failure, `expectNoMsg(fixedDelay)` is set 1 second, the actual delay time is 1045ms, so why the test fails on `expectNoMsg`?

**Comment by manonthegithub** [comment]
Isn't it a bug in TestSubscriber Probe and what size of the gap to avoid inaccuracy should be?

**Comment by manonthegithub** [comment]
The notable point is that, now I reproduced it and it failed on 1011ms despite the fact that maximun delay in a row was 1022ms.

**Comment by johanandren** [comment]
And those you measured with `nanoTime`, not `currentTimeMillis`?

**Comment by manonthegithub** [comment]
yes
image

on the bottom row test failed

**Comment by manonthegithub** [comment]
I've got more samples, there were 705 successful passes. The summary is (diff time in nanos):
image

**Comment by manonthegithub** [comment]
I've done more research.
I decided to write test without delay stage to test `expectNoMsg`:
```scala
"test" in {
      for(_ <- 1 to 10000) {
        val blockingDispatcher = ExecutionContext.fromExecutor(Executors.newSingleThreadExecutor())

        val elems = Vector(1,2)

        val probe = Source(elems)
          .log("l1", e => println(s"Test in $e: ${System.nanoTime()}"))
          .mapAsync(1)(e => Future{Thread.sleep(1010); e}(blockingDispatcher))//1007 fails
          .log("l2", e => println(s"Test out $e: ${System.nanoTime()}"))
          .runWith(TestSink.probe)

        elems.foreach(_ should be(probe
          .request(1)
          .expectNoMsg(1 second)
          .expectNext()))

        probe
          .request(1)
          .expectNextOrComplete(1)

      }
    }
```
If I set `Thread.sleep` to `1007` or less it fails almost immediately. Setting it to `1010` made it to work for few hours and then it failed too with:
```
assertion failed: received unexpected message OnNext(1)
java.lang.AssertionError: assertion failed: received unexpected message OnNext(1)
	at scala.Predef$.assert(Predef.scala:219)
	at akka.testkit.TestKitBase.expectNoMsg_internal(TestKit.scala:616)
```
The measurements show that failure was on `1010481768` nanos.
WDYT @johanandren ?

**Comment by manonthegithub** [comment]
For now I completely removed expectNoMsg from tests and replaces it with System.nanoTime calculations

**Comment by manonthegithub** [comment]
Guys, there are some heavy delays on build servers, that fail tests. I can increase wait time in my tests but there are some more tests failing on timeout that are not dependent on this PR. What do you think is better solution for this?

**Comment by raboof** [comment]
@manonthegithub I agree this is getting problematic. While travis can be somewhat jittery, I think we should improve our tests so that they're more resilient.

Feel free to create issues for specific unreliable tests, we'll use the 'failure' label to track them. Fixing those should be a priority.

I'm happy to restart builds after we verified that there's issues for the failures, but it'd be really nice to at least have the `DelayFlow` tests stable before merging this PR...

**Comment by manonthegithub** [comment]
@raboof
I agree, but my tests largely depend on timing. If everything gets slow and messages get delayed in an unpredictable way, it is problematic to fix them, the only thing I can do is to increase wait time then, which becomes much longer than expected... So these tests will have the only guarantee that elems get delayed by `DelayFlow` for 'at least' the specified delay but this delay can last longer one would expect.

**Comment by manonthegithub** [comment]
After increased delay `DelayFlow` tests passed. Can you please retrigger it several times for sure?

**Comment by manonthegithub** [comment]
squashed and rebased

**Comment by manonthegithub** [comment]
I guess, It can be merged, because there are no tests in my PR that are failing (I also increased timeout in them). The failed ones are #109 and #104. I you will have have time, please review latest version).

**Comment by manonthegithub** [comment]
Great. Thanks!

**Pull request merged**